### PR TITLE
Generic class model for Java, Kotlin and Scala

### PR DESCRIPTION
**Pull request by wilmveel**
This PR introduces an extra step in the emit process for class base languages.
Goal: the emitter should have a single responsibility.
Currently the emitter is transforming the structure an outputting the code. Instead of emitting the AST direct there is an intermediate model which does the transformation. So the emitter is only responsible for outputting the code.

AST -> ClassModel -> EmitLanguage

@jerrevanveluw sanitize is stil part of the emit process ...

### CODE REVIEW

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt [hunk 1]
```diff
@@ -55,240 +57,195 @@ class KotlinEmitter(
...
-    }
 
-    override fun Node.emitName(): String = when (this) {
-        is Endpoint -> "${this.name}Endpoint"
-        is Enum -> this.name
-        is Refined -> this.name
-        is Type -> this.name
-    }
+    override fun EndpointClass.Content.emit(): String = """
+        |Wirespec.Content("$type", body)
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt:156]: Could be without `trimMargin()`

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt [hunk 2]
```diff
@@ -55,240 +57,195 @@ class KotlinEmitter(
...
-            |(${
-            parameters
-                .plus(content?.reference?.toField("body", false))
-                .filterNotNull()
-                .joinToString(", ") { it.emit() }
-        })
-        """.trimMargin()
-    }
+    override fun EndpointClass.RequestMapper.emit(): String = """
+        |fun <B> $name(contentMapper: Wirespec.ContentMapper<B>) = { request: Wirespec.Request<B> ->
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt:160]: Guess this needs a.space :P

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/CompileEnumTest.kt [hunk 3]
```diff
@@ -25,47 +27,50 @@ class CompileEnumTest {
...
+            |  }
+            |  @Override
+            |  public String toString() {
+            |    return label;
+            |  }
+            |}
+            |
+        """.trimMargin()
+
+        assertEquals(java, compiler(JavaEmitter(logger = logger)).getOrNull())
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/CompileEnumTest.kt:73]: assertion can be removed

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitterTest.kt [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+            |  }
+            |  @Override
+            |  public String toString() {
+            |    return label;
+            |  }
+            |}
+        """.trimMargin()
+
+        val res = emitter.emit(ClassModelFixture.enum)
+        assertEquals(expected, res)
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitterTest.kt:62]: res shouldBe expected

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitterTest.kt [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+            |    };
+            |  }
+            |
+            |  public CompletableFuture<Response<?>> addPet(Request<?> request);
+            |
+            |}
+        """.trimMargin()
+
+        val res = emitter.emit(ClassModelFixture.endpoint)
+        assertEquals(expected, res)
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitterTest.kt:383]: res shouldBe expected

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitterTest.kt [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            |data class Todo(
+            |  val name: String,
+            |  val description: String? = null,
+            |  val notes: List<String>,
+            |  val done: Boolean
+            |)
+        """.trimMargin()
+
+        val res = emitter.emit(ClassModelFixture.type)
+        assertEquals(expected, res)
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitterTest.kt:23]: res shouldBe expected

**Review by jerrevanveluw — CHANGES_REQUESTED** [review]

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt [hunk 7]
```diff
@@ -53,272 +59,294 @@ class JavaEmitter(
...
-            |) {};
-            |""".trimMargin()
-    }
+    override fun emit(ast: List<Definition>): List<Emitted> = super.emit(ast)
+        .map { Emitted(it.typeName.sanitizeSymbol(), "$pkg${importWireSpec(ast)}${importJava(ast)}${it.result}\n") }
 
-    override fun Type.Shape.emit() = withLogging(logger) {
-        value.joinToString(",\n") { it.emit() }
-    }
+    override fun TypeClass.emit(): String = """
```
[the remark this answers is not in the document]
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:91]: Dit is iets anders. Ander domain.

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt [hunk 8]
```diff
@@ -53,272 +59,294 @@ class JavaEmitter(
...
+        .flatMap { listOf(EndpointClass.Path.Literal("/"), it) }
+        .joinToString(" + ") {
+            when (it) {
+                is EndpointClass.Path.Literal -> "\"${it.value}\""
+                is EndpointClass.Path.Parameter -> it.value
+            }
+        }
+
+    override fun EndpointClass.Content.emit(): String = """
+        |new Wirespec.Content("$type", body)
```
[the remark this answers is not in the document]
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:316]: Without `trimMargin` you have to escape the `\"` in don't like that

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt [hunk 9]
```diff
@@ -53,272 +59,294 @@ class JavaEmitter(
...
+    """.trimMargin()
+
+    override fun Field.emit(): String = """
+        |${if (isPrivate) "private " else ""}${if (isPrivate) "final " else ""}${reference.emitWrap()} ${identifier.sanitizeIdentifier()}
```
[the remark this answers is not in the document]
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:320]: `\"`

[the remark this answers is not in the document]
Comment (jerrevanveluw) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:319]: Nope ;)

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt (hunk 2 above)
Comment (jerrevanveluw) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt:160]: In front of the `|` it is not in line with the next line

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt [hunk 10]
```diff
@@ -55,240 +57,196 @@ class KotlinEmitter(
...
+         |${SPACER}}
+         |}
+    """.trimMargin()
 
-    private fun List<Type.Shape.Field>.emitMap() =
-        joinToString(", ") { "\"${it.identifier.value}\" to listOf(${it.identifier.emit()})" }
+    override fun EndpointClass.RequestMapper.RequestCondition.emit(): String =
+        if (content == null)
+            """
+                |request.content == null -> ${responseReference.emitWrap()}(request.path, request.method, request.query, request.headers)
```
Comment (jerrevanveluw) [inline src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt:171]: """request.content == null -> ${responseReference.emitWrap()}(request.path, request.method, request.query, request.headers)"""

**Review by jerrevanveluw — CHANGES_REQUESTED** [review]

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:90]: Het is nu consistent met <WsKeyword>Type. Ik zou dat graag behouden.

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt (hunk 2 above)
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/KotlinEmitter.kt:160]: Check

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (wilmveel) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/emit/JavaEmitter.kt:319]: fixed

**Review by jerrevanveluw — APPROVED** [review]

**Pull request merged**